### PR TITLE
Enable publishing to build storage

### PR DESCRIPTION
**Pull request by NikolaMilosavljevic**
Contributes to https://github.com/dotnet/source-build/issues/4101

Enables publishing to local build storage. It will be used in VMR and requires a new optional task parameter.

Some highlights:
- A bit of refactoring of some of `arcade` projects to enable building in source-only build
- Removal of NuGet.Packaging dependency and implementation of required APIs
- Exclusion of a set of sources in `Microsoft.DotNet.Build.Tasks.Feed` project
- Renaming of `PushToAzureDevOpsArtifacts` task to `PushToBuildStorage`

I've verified that regular repo build and source-build are working fine with these changes. I have also fully verified Windows and Linux VMR build Additional PRs for `installer` and other repos will follow soon.

Full set of VMR changes for my testing can be seen here: https://github.com/dotnet/dotnet/compare/d3e9fdb1..9ac3fddf

### CODE REVIEW

File: src/Microsoft.DotNet.VersionTools/tools/Microsoft.DotNet.VersionTools.Cli/Microsoft.DotNet.VersionTools.Cli.csproj [hunk 1]
```diff
@@ -9,6 +9,12 @@
     <Description>This package provides a CLI interface to the Microsoft.DotNet.VersionTools library.</Description>
   </PropertyGroup>
 
+  <!-- Don't include VersionTools.Cli in source-build until we find a reason to use it. -->
+  <PropertyGroup>
+    <ExcludeFromSourceBuild>true</ExcludeFromSourceBuild>
+    <ExcludeFromSourceOnlyBuild>true</ExcludeFromSourceOnlyBuild>
+  </PropertyGroup>
```
Comment (ViktorHofer) [inline src/Microsoft.DotNet.VersionTools/tools/Microsoft.DotNet.VersionTools.Cli/Microsoft.DotNet.VersionTools.Cli.csproj:16]: You only need the new switch.

Was there a reason to not build this project? I think I would opt towards building as much as possible in source-build to reduce the build diff.

File: src/Microsoft.DotNet.VersionTools/tools/Microsoft.DotNet.VersionTools.Cli.Tests/Microsoft.DotNet.VersionTools.Cli.Tests.csproj [hunk 2]
```diff
@@ -4,6 +4,12 @@
     <TargetFramework>$(NetToolCurrent)</TargetFramework>
   </PropertyGroup>
 
+  <!-- Don't include VersionTools.Cli.Tests in source-build until we find a reason to use it. -->
+  <PropertyGroup>
+    <ExcludeFromSourceBuild>true</ExcludeFromSourceBuild>
+    <ExcludeFromSourceOnlyBuild>true</ExcludeFromSourceOnlyBuild>
+  </PropertyGroup>
+
```
Comment (ViktorHofer) [inline src/Microsoft.DotNet.VersionTools/tools/Microsoft.DotNet.VersionTools.Cli.Tests/Microsoft.DotNet.VersionTools.Cli.Tests.csproj:12]: Test projects are already excluded by default, you can remove this.

File: src/Microsoft.DotNet.VersionTools/tests/Microsoft.DotNet.VersionTools.Tests.csproj [hunk 3]
```diff
@@ -4,6 +4,12 @@
     <TargetFramework>$(NetToolCurrent)</TargetFramework>
   </PropertyGroup>
 
+  <!-- Don't include VersionTools.Tests in source-build until we find a reason to use it. -->
+  <PropertyGroup>
+    <ExcludeFromSourceBuild>true</ExcludeFromSourceBuild>
+    <ExcludeFromSourceOnlyBuild>true</ExcludeFromSourceOnlyBuild>
+  </PropertyGroup>
+
```
Comment (ViktorHofer) [inline src/Microsoft.DotNet.VersionTools/tests/Microsoft.DotNet.VersionTools.Tests.csproj:12]: Same here. Test projects are already excluded by default.

File: src/Microsoft.DotNet.VersionTools/tasks/Microsoft.DotNet.VersionTools.Tasks.csproj [hunk 4]
```diff
@@ -7,6 +7,12 @@
     <PackageType>MSBuildSdk</PackageType>
   </PropertyGroup>
 
+  <!-- Don't include VersionTools.Tasks in source-build until we find a reason to use it. -->
+  <PropertyGroup>
+    <ExcludeFromSourceBuild>true</ExcludeFromSourceBuild>
+    <ExcludeFromSourceOnlyBuild>true</ExcludeFromSourceOnlyBuild>
+  </PropertyGroup>
+
```
Comment (ViktorHofer) [inline src/Microsoft.DotNet.VersionTools/tasks/Microsoft.DotNet.VersionTools.Tasks.csproj:15]: Same question as for the Cli frontend. Does building this cause any prebuilt issues? If not, I would prefer if we would build this project in source-build to reduce the build diff.

**Review by ViktorHofer — COMMENTED** [review]
You are probably aware of this but just to mention it here, this will need reaction in multiple repos that currently call this task directly: https://github.com/search?q=org%3Adotnet%20PushToAzureDevOpsArtifacts&type=code

File: src/Microsoft.DotNet.VersionTools/tools/Microsoft.DotNet.VersionTools.Cli/Microsoft.DotNet.VersionTools.Cli.csproj (hunk 1 above)
Comment (NikolaMilosavljevic) [reply src/Microsoft.DotNet.VersionTools/tools/Microsoft.DotNet.VersionTools.Cli/Microsoft.DotNet.VersionTools.Cli.csproj:16]: Building this would introduce a dozen of prebuilts. The project was already excluded before at the higher level dir.

File: src/Microsoft.DotNet.VersionTools/tasks/Microsoft.DotNet.VersionTools.Tasks.csproj (hunk 4 above)
Comment (NikolaMilosavljevic) [reply src/Microsoft.DotNet.VersionTools/tasks/Microsoft.DotNet.VersionTools.Tasks.csproj:15]: Building this would introduce a dozen of prebuilts. The project was already excluded before at the higher level dir.

File: src/Microsoft.DotNet.Build.Tasks.Feed/Microsoft.DotNet.Build.Tasks.Feed.csproj [hunk 5]
```diff
@@ -7,17 +7,19 @@
...
+    <PackageReference Include="Azure.Core" Condition="'$(DotNetBuildSourceOnly)' != 'true'" />
+    <PackageReference Include="Azure.Storage.Blobs" Condition="'$(DotNetBuildSourceOnly)' != 'true'" />
     <PackageReference Include="Microsoft.Build.Tasks.Core" />
 
     <PackageReference Include="Microsoft.Extensions.DependencyInjection.Abstractions" />
-    <PackageReference Include="Microsoft.SymbolUploader" />
+    <PackageReference Include="Microsoft.SymbolUploader" Condition="'$(DotNetBuildSourceOnly)' != 'true'" />
   </ItemGroup>
 
   <!-- Upgrade SymbolUploader's transitive NETStandard.Library dependency to avoid .NET Standard 1.x dependencies. -->
```
Comment (ViktorHofer) [inline src/Microsoft.DotNet.Build.Tasks.Feed/Microsoft.DotNet.Build.Tasks.Feed.csproj:25]: Condiiton the NETStandard.Library package refernce on DotNetBuildSourceOnly != true as well.

**Comment by NikolaMilosavljevic** [comment]
> You are probably aware of this but just to mention it here, this will need reaction in multiple repos that currently call this task directly: https://github.com/search?q=org%3Adotnet%20PushToAzureDevOpsArtifacts&type=code

Yes, we've discussed this in https://github.com/dotnet/source-build/issues/4101. We'll need to update those references when new `arcade` flows to those repos.

File: src/Microsoft.DotNet.VersionTools/tools/Microsoft.DotNet.VersionTools.Cli/Microsoft.DotNet.VersionTools.Cli.csproj (hunk 1 above)
Comment (ViktorHofer) [reply src/Microsoft.DotNet.VersionTools/tools/Microsoft.DotNet.VersionTools.Cli/Microsoft.DotNet.VersionTools.Cli.csproj:16]: Which prebuilts? I only see it referencing System.CommandLine which is available in source-build.

**Comment by MilenaHristova** [comment]
@NikolaMilosavljevic I recently added `DotNetReleaseShipping=true` metadata in the manifest for assets that get shipped with the release infra, is this preserved with your change?

Here is the issue: https://github.com/dotnet/release/issues/785 - I added a property `ProducesDotNetReleaseShippingAssets` in  ([Publishing.props](https://github.com/dotnet/emsdk/blob/a5f4de78fca42544771977f8e8e04c4aa83e1d02/eng/Publishing.props#L6C8-L6C43)) in the shipping repos and it's getting propagated to the assets manifest in [Publish.proj](https://github.com/dotnet/arcade/blob/2215d18d04a6565e9133c15ece7df7161fdf1e8e/src/Microsoft.DotNet.Arcade.Sdk/tools/Publish.proj#L111). It should be set for all shipping assets and packages produced from the repos that have `ProducesDotNetReleaseShippingAssets` set to true.

File: src/Microsoft.DotNet.VersionTools/tools/Microsoft.DotNet.VersionTools.Cli/Microsoft.DotNet.VersionTools.Cli.csproj (hunk 1 above)
Comment (NikolaMilosavljevic) [reply src/Microsoft.DotNet.VersionTools/tools/Microsoft.DotNet.VersionTools.Cli/Microsoft.DotNet.VersionTools.Cli.csproj:16]: I'm going to keep both properties. This essentially moves the existing infra from https://github.com/dotnet/arcade/blob/2215d18d04a6565e9133c15ece7df7161fdf1e8e/src/Microsoft.DotNet.VersionTools/Directory.Build.props#L5-L9 into project file.

Removing `ExcludeFromSourceBuild` property would likely require building more of the other arcade projects, that this one depends on, that are currently disabled with `ExcludeFromSourceBuild`.

I believe the comment is still true - we do not enable this unless we find a reason to use it.

File: src/Microsoft.DotNet.VersionTools/tasks/Microsoft.DotNet.VersionTools.Tasks.csproj (hunk 4 above)
Comment (NikolaMilosavljevic) [reply src/Microsoft.DotNet.VersionTools/tasks/Microsoft.DotNet.VersionTools.Tasks.csproj:15]: Will keep this. Same comment as for VersionTools.Cli project above.

File: src/Microsoft.DotNet.VersionTools/tools/Microsoft.DotNet.VersionTools.Cli/Microsoft.DotNet.VersionTools.Cli.csproj (hunk 1 above)
Comment (ViktorHofer) [reply src/Microsoft.DotNet.VersionTools/tools/Microsoft.DotNet.VersionTools.Cli/Microsoft.DotNet.VersionTools.Cli.csproj:16]: ExcludeFromSourceBuild is unnecessary in Arcade's own build. The old switch is only required in repositories that still use Arcade 8. Arcade main depends on Arcade 9 so it doesn't need the old switch anymore. I'm happy to submit a separate PR if you rather not want to include that change in your PR.

> I believe the comment is still true - we do not enable this unless we find a reason to use it.

This doesn't make sense to me. Nowhere else in the stack do we disable projects when building from source unless they would bring in prebuilts or add cost to the build, i.e. VS insertion packages, workloads, etc. Reducing the build difference between source-build and msft-build has been a goal for years by your team so I'm surprised that we aren't in agreement here.

File: src/Microsoft.DotNet.VersionTools/tools/Microsoft.DotNet.VersionTools.Cli.Tests/Microsoft.DotNet.VersionTools.Cli.Tests.csproj (hunk 2 above)
Comment (mmitche) [reply src/Microsoft.DotNet.VersionTools/tools/Microsoft.DotNet.VersionTools.Cli.Tests/Microsoft.DotNet.VersionTools.Cli.Tests.csproj:12]: Yep. And anywhwere you have both `<ExcludeFromSourceOnlyBuild>true</ExcludeFromSourceOnlyBuild>` and `<ExcludeFromSourceBuild>true</ExcludeFromSourceBuild>` you can keep only the former. The only reason that both still exist in arcade is that repos that insert into the 8.0 SDK use the old switches and so it still needs to support them. But this doesn't apply to arcade's own internal usage.

**Comment by NikolaMilosavljevic** [comment]
> DotNetReleaseShipping

Yes, the metadata exists in several repo manifests - `aspnetcore`, `emsdk`, `roslyn-analyzers`, `runtime`, `sdk` and `templating`.

File: src/Microsoft.DotNet.VersionTools/tools/Microsoft.DotNet.VersionTools.Cli.Tests/Microsoft.DotNet.VersionTools.Cli.Tests.csproj (hunk 2 above)
Comment (NikolaMilosavljevic) [reply src/Microsoft.DotNet.VersionTools/tools/Microsoft.DotNet.VersionTools.Cli.Tests/Microsoft.DotNet.VersionTools.Cli.Tests.csproj:12]: I see. So I can just keep `ExcludeFromSourceOnlyBuild` and it will not start building the project in any VMR build leg, right?

Comment (mmitche) [reply src/Microsoft.DotNet.VersionTools/tools/Microsoft.DotNet.VersionTools.Cli.Tests/Microsoft.DotNet.VersionTools.Cli.Tests.csproj:12]: Correct

File: src/Microsoft.DotNet.VersionTools/tools/Microsoft.DotNet.VersionTools.Cli/Microsoft.DotNet.VersionTools.Cli.csproj (hunk 1 above)
Comment (ViktorHofer) [reply src/Microsoft.DotNet.VersionTools/tools/Microsoft.DotNet.VersionTools.Cli/Microsoft.DotNet.VersionTools.Cli.csproj:16]: No worries, https://github.com/dotnet/arcade/pull/14564 will take care of those.

Comment (NikolaMilosavljevic) [reply src/Microsoft.DotNet.VersionTools/tools/Microsoft.DotNet.VersionTools.Cli/Microsoft.DotNet.VersionTools.Cli.csproj]: Fixed with https://github.com/dotnet/arcade/pull/14559/commits/913dbfbfed72e9b68a10851456fe49571a11839f

File: src/Microsoft.DotNet.VersionTools/tools/Microsoft.DotNet.VersionTools.Cli.Tests/Microsoft.DotNet.VersionTools.Cli.Tests.csproj (hunk 2 above)
Comment (NikolaMilosavljevic) [reply src/Microsoft.DotNet.VersionTools/tools/Microsoft.DotNet.VersionTools.Cli.Tests/Microsoft.DotNet.VersionTools.Cli.Tests.csproj]: Fixed with https://github.com/dotnet/arcade/pull/14559/commits/913dbfbfed72e9b68a10851456fe49571a11839f

File: src/Microsoft.DotNet.VersionTools/tests/Microsoft.DotNet.VersionTools.Tests.csproj (hunk 3 above)
Comment (NikolaMilosavljevic) [reply src/Microsoft.DotNet.VersionTools/tests/Microsoft.DotNet.VersionTools.Tests.csproj]: Fixed with https://github.com/dotnet/arcade/pull/14559/commits/913dbfbfed72e9b68a10851456fe49571a11839f

File: src/Microsoft.DotNet.VersionTools/tasks/Microsoft.DotNet.VersionTools.Tasks.csproj (hunk 4 above)
Comment (NikolaMilosavljevic) [reply src/Microsoft.DotNet.VersionTools/tasks/Microsoft.DotNet.VersionTools.Tasks.csproj]: Fixed with https://github.com/dotnet/arcade/pull/14559/commits/913dbfbfed72e9b68a10851456fe49571a11839f

File: src/Microsoft.DotNet.Build.Tasks.Feed/Microsoft.DotNet.Build.Tasks.Feed.csproj (hunk 5 above)
Comment (NikolaMilosavljevic) [reply src/Microsoft.DotNet.Build.Tasks.Feed/Microsoft.DotNet.Build.Tasks.Feed.csproj:25]: Fixed with https://github.com/dotnet/arcade/pull/14559/commits/913dbfbfed72e9b68a10851456fe49571a11839f

File: src/Microsoft.DotNet.VersionTools/tools/Microsoft.DotNet.VersionTools.Cli/Microsoft.DotNet.VersionTools.Cli.csproj (hunk 1 above)
Comment (ViktorHofer) [reply src/Microsoft.DotNet.VersionTools/tools/Microsoft.DotNet.VersionTools.Cli/Microsoft.DotNet.VersionTools.Cli.csproj]: My point still stands that we whould build the Cli and Task projects as those don't introduce new prebuilts do minimize the difference between the msft-build and source-build but I'm fine if you do that as a follow-up after this change is in to reduce the build diff in this PR.

File: src/Microsoft.DotNet.VersionTools/Directory.Build.props [hunk 6]
```diff
@@ -2,9 +2,4 @@
 
   <Import Project="$([MSBuild]::GetDirectoryNameOfFileAbove($(MSBuildThisFileDirectory)\.., Directory.Build.props))\Directory.Build.props" />
 
-  <!-- Don't include VersionTools in source-build until we find a reason to use it. -->
```
Comment (ViktorHofer) [inline src/Microsoft.DotNet.VersionTools/Directory.Build.props:5]: Delete this file now that it doesn't do anything anymore.

File: src/Microsoft.DotNet.VersionTools/lib/src/Automation/NupkgInfo.cs [hunk 7]
```diff
@@ -10,14 +10,40 @@ public class NupkgInfo
...
             Id = identity.Id;
-            Version = identity.Version.ToString();
-            Prerelease = identity.Version.Release;
+            Version = identity.Version;
         }
 
         public string Id { get; }
         public string Version { get; }
-        public string Prerelease { get; }
+        public string Prerelease { get { throw new NotImplementedException();} }
```
Comment (ViktorHofer) [inline src/Microsoft.DotNet.VersionTools/lib/src/Automation/NupkgInfo.cs:18]: Why can't you remove the API?

File: src/Microsoft.DotNet.VersionTools/lib/src/Automation/NupkgInfo.cs [hunk 8]
```diff
@@ -10,14 +10,40 @@ public class NupkgInfo
...
+
+        public string Id
+        {
+            get { return _id; }
+        }
+
+        public string Version
+        {
+            get { return _version; }
+        }
```
Comment (ViktorHofer) [inline src/Microsoft.DotNet.VersionTools/lib/src/Automation/NupkgInfo.cs:47]:
```suggestion
        public string Id { get; init; }
        public string Version { get; init; }

        public PackageIdentity(string id, string version)
        {
            if (id == null)
            {
                throw new ArgumentNullException(nameof(id));
            }

            Id = id;
            Version = version;
        }
```

File: src/Microsoft.DotNet.VersionTools/lib/src/Automation/NupkgInfoFactory.cs [hunk 9]
```diff
@@ -22,10 +26,61 @@ public NupkgInfoFactory(IPackageArchiveReaderFactory packageArchiveReaderFactory
...
+                stream = File.OpenRead(path);
+                ZipReadStream = stream;
+                _zipArchive = new ZipArchive(stream, ZipArchiveMode.Read);
+                string nuspecFile = Path.GetTempFileName();
+                foreach (ZipArchiveEntry entry in _zipArchive.Entries)
+                {
+                    if (entry.Name.EndsWith(".nuspec"))
+                    {
+                        Directory.CreateDirectory(Path.GetDirectoryName(nuspecFile));
+                        entry.ExtractToFile(nuspecFile, true);
```
Comment (ViktorHofer) [inline src/Microsoft.DotNet.VersionTools/lib/src/Automation/NupkgInfoFactory.cs:48]: This should be possible in-memory without writing to disk by using the `Open` API.

File: src/Microsoft.DotNet.VersionTools/lib/src/Automation/NupkgInfoFactory.cs [hunk 10]
```diff
@@ -22,10 +26,61 @@ public NupkgInfoFactory(IPackageArchiveReaderFactory packageArchiveReaderFactory
...
+            XDocument doc = XDocument.Load(nuspecFile, LoadOptions.PreserveWhitespace);
+            XElement metadataElement = GetSingleElement(doc.Root, "metadata");
+            return new PackageIdentity(GetSingleElement(metadataElement, "id").Value, GetSingleElement(metadataElement, "version").Value);
+        }   
+
+        private static XElement GetSingleElement(XElement el, string name)
+        {
+            return el.Descendants()
+                .Where(c => c.Name.LocalName.ToString() == name)
+                .ToArray().First();
```
Comment (ViktorHofer) [inline src/Microsoft.DotNet.VersionTools/lib/src/Automation/NupkgInfoFactory.cs:83]: Instead of using Where() + First(), just use `.First(c => c.Name.LocalName.ToString() == name)`.

File: src/Microsoft.DotNet.Arcade.Sdk/tools/Publish.proj [hunk 11]
```diff
@@ -85,6 +86,13 @@
         <IsShipping>%(PackagesToPublish.IsShipping)</IsShipping>
       </SymbolPackagesToGenerate>
 
+      <!-- Include Symbols.<repo>.tar.gz, if running in inner source-only build -->
+      <UnifiedSymbolsPackage Include="$(ArtifactsPackagesDir)**/Symbols*.tar.gz" IsShipping="false" />
```
Comment (ViktorHofer) [inline src/Microsoft.DotNet.Arcade.Sdk/tools/Publish.proj:90]: Do we need the glob here, can this path be more specific?

Comment (ViktorHofer) [inline src/Microsoft.DotNet.Arcade.Sdk/tools/Publish.proj:90]: This item should also have the same condition as the item below (only for source-build). I would group them together into an own item group with the condition for the whole group.

File: src/Microsoft.DotNet.Arcade.Sdk/tools/Publish.proj [hunk 12]
```diff
@@ -145,6 +156,11 @@
       <PDBsToPublishTempLocation>$(ArtifactsTmpDir)PDBsToPublish/</PDBsToPublishTempLocation>
     </PropertyGroup>
     
+    <!-- In VMR inner-build copy to local storage, do not push to AzDO -->
+    <PropertyGroup Condition="'$(ArcadeInnerBuildFromSource)' == 'true' or '$(DotNetBuildPhase)' == 'InnerRepo'">
+      <PushToLocalStorage>true</PushToLocalStorage>
+    </PropertyGroup>
```
Comment (ViktorHofer) [inline src/Microsoft.DotNet.Arcade.Sdk/tools/Publish.proj:162]: Move this up after L23 and before L24 so that Publishing.props logic can use this property for their conditions.

File: src/Microsoft.DotNet.Arcade.Sdk/tools/SourceBuild/SourceBuildArcade.targets [hunk 13]
```diff
@@ -159,12 +160,12 @@
...
   <Target Name="CreateRepoSymbolsArchive"
           Condition="'$(OS)' != 'Windows_NT' and ('$(DotNetBuildFromSourceFlavor)' == 'Product' or '$(DotNetBuildOrchestrator)' == 'true' or '$(SupplementalIntermediateNupkgCategory)' == '$(SymbolsIntermediateNupkgCategory)')"
-          DependsOnTargets="GetCategorizedIntermediateNupkgContents">
+          DependsOnTargets="$(CreateRepoSymbolsArchiveDependsOn)">
     <PropertyGroup>
       <SymbolsRoot>$(CurrentRepoSourceBuildArtifactsDir)</SymbolsRoot>
       <!-- Fall back to repo root for source-build-externals or repos that don't have the regular SymbolsRoot as defined above -->
       <SymbolsRoot Condition="!Exists('$(SymbolsRoot)') or '$(GitHubRepositoryName)' == 'source-build-externals'">$(RepoRoot)</SymbolsRoot>
-      <SymbolsArchiveLocation>$(CurrentRepoSourceBuildArtifactsPackagesDir)</SymbolsArchiveLocation>
+      <SymbolsArchiveLocation>$(CurrentRepoSourceBuildArtifactsPackagesDir)NonShipping/</SymbolsArchiveLocation>
```
Comment (ViktorHofer) [inline src/Microsoft.DotNet.Arcade.Sdk/tools/SourceBuild/SourceBuildArcade.targets:168]: Please create a separate property for the non shipping packages folder after L48 and use the NormalizeDirectory function.

**Comment by ViktorHofer** [comment]
@mmitche I did a review of most of the msbuild stuff but I only glanced at the publishing C# code. Can you please also take look?

File: src/Microsoft.DotNet.VersionTools/lib/src/Automation/NupkgInfo.cs (hunk 7 above)
Comment (NikolaMilosavljevic) [reply src/Microsoft.DotNet.VersionTools/lib/src/Automation/NupkgInfo.cs:18]: This was suggested by @mmitche in case there are some unexpected references of this API somewhere else.

File: src/Microsoft.DotNet.VersionTools/Directory.Build.props (hunk 6 above)
Comment (NikolaMilosavljevic) [reply src/Microsoft.DotNet.VersionTools/Directory.Build.props:5]: Right - good call.

File: src/Microsoft.DotNet.Arcade.Sdk/tools/Publish.proj (hunk 11 above)
Comment (NikolaMilosavljevic) [reply src/Microsoft.DotNet.Arcade.Sdk/tools/Publish.proj:90]: Right - this can be more specific - will fix.

File: src/Microsoft.DotNet.VersionTools/lib/src/Automation/NupkgInfoFactory.cs (hunk 10 above)
Comment (NikolaMilosavljevic) [reply src/Microsoft.DotNet.VersionTools/lib/src/Automation/NupkgInfoFactory.cs:83]: Good call - thanks!

File: src/Microsoft.DotNet.VersionTools/lib/src/Automation/NupkgInfo.cs (hunk 8 above)
Comment (NikolaMilosavljevic) [reply src/Microsoft.DotNet.VersionTools/lib/src/Automation/NupkgInfo.cs]: Interestingly, this did not compile in VMR build. Of course, it works fine in simple VS projects.

I will revert this change when I push the next set of changes to address other comments. Old code was totally fine anyway, and is used by NuGet.Client as well: https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Core/NuGet.Packaging/Core/PackageIdentity.cs

Comment (ViktorHofer) [reply src/Microsoft.DotNet.VersionTools/lib/src/Automation/NupkgInfo.cs]: Interesting. I will play around with this post-merge.

File: src/Microsoft.DotNet.VersionTools/Directory.Build.props (hunk 6 above)
Comment (NikolaMilosavljevic) [reply src/Microsoft.DotNet.VersionTools/Directory.Build.props:5]: Fixed with https://github.com/dotnet/arcade/pull/14559/commits/4cce0e9f7c41c3a832d809c1d71cb3380f99a6ba

File: src/Microsoft.DotNet.VersionTools/lib/src/Automation/NupkgInfoFactory.cs (hunk 10 above)
Comment (NikolaMilosavljevic) [reply src/Microsoft.DotNet.VersionTools/lib/src/Automation/NupkgInfoFactory.cs]: Fixed with https://github.com/dotnet/arcade/pull/14559/commits/4cce0e9f7c41c3a832d809c1d71cb3380f99a6ba

File: src/Microsoft.DotNet.Arcade.Sdk/tools/Publish.proj (hunk 11 above)
Comment (NikolaMilosavljevic) [reply src/Microsoft.DotNet.Arcade.Sdk/tools/Publish.proj]: Fixed with https://github.com/dotnet/arcade/pull/14559/commits/4cce0e9f7c41c3a832d809c1d71cb3380f99a6ba

Comment (NikolaMilosavljevic) [reply src/Microsoft.DotNet.Arcade.Sdk/tools/Publish.proj]: Fixed with https://github.com/dotnet/arcade/pull/14559/commits/4cce0e9f7c41c3a832d809c1d71cb3380f99a6ba

File: src/Microsoft.DotNet.Arcade.Sdk/tools/Publish.proj (hunk 12 above)
Comment (NikolaMilosavljevic) [reply src/Microsoft.DotNet.Arcade.Sdk/tools/Publish.proj]: Fixed with https://github.com/dotnet/arcade/pull/14559/commits/4cce0e9f7c41c3a832d809c1d71cb3380f99a6ba

File: src/Microsoft.DotNet.Arcade.Sdk/tools/SourceBuild/SourceBuildArcade.targets (hunk 13 above)
Comment (NikolaMilosavljevic) [reply src/Microsoft.DotNet.Arcade.Sdk/tools/SourceBuild/SourceBuildArcade.targets]: Fixed with https://github.com/dotnet/arcade/pull/14559/commits/4cce0e9f7c41c3a832d809c1d71cb3380f99a6ba

File: src/Microsoft.DotNet.Arcade.Sdk/tools/SourceBuild/SourceBuildArcadeBuild.targets [hunk 14]
```diff
@@ -80,6 +80,9 @@
       <InnerBuildArgs Condition="'$(DotNetPackageVersionPropsPath)' != ''">$(InnerBuildArgs) /p:DotNetPackageVersionPropsPath="$(DotNetPackageVersionPropsPath)"</InnerBuildArgs>
 
       <InnerBuildArgs>$(InnerBuildArgs) /p:FullAssemblySigningSupported=$(FullAssemblySigningSupported)</InnerBuildArgs>
+
+      <!-- We need to utilize publishing to copy packages, assets and manifest to shared source-build location. -->
+      <InnerBuildArgs>$(InnerBuildArgs) /p:DotNetPublishUsingPipelines=true</InnerBuildArgs>
```
Comment (ViktorHofer) [inline src/Microsoft.DotNet.Arcade.Sdk/tools/SourceBuild/SourceBuildArcadeBuild.targets:85]: Just to double check, @mmitche is this right?

File: src/Microsoft.DotNet.Arcade.Sdk/tools/Publish.proj [hunk 15]
```diff
@@ -21,6 +21,15 @@
   -->
 
   <Import Project="BuildStep.props" />
+
+  <!--
+    In VMR inner-build copy to local storage, do not push to AzDO.
+    Define this property before importing repo Publishing.props, to have it control any repo-specific publishing.
+  -->
+  <PropertyGroup Condition="'$(DotNetBuildInnerRepo)' == 'true' and '$(DotNetBuildOrchestrator)' == 'true'">
```
Comment (ViktorHofer) [inline src/Microsoft.DotNet.Arcade.Sdk/tools/Publish.proj:29]: If we would utilize the Publish.proj at the end of the VMR build, this would be wrong then as we then would want to publish to remote.

Can we condition this on an exclusive property, so only if we are inside the repo inner build?

Comment (ViktorHofer) [reply src/Microsoft.DotNet.Arcade.Sdk/tools/Publish.proj:29]: cc @mmitche

Comment (NikolaMilosavljevic) [reply src/Microsoft.DotNet.Arcade.Sdk/tools/Publish.proj:29]: `DotNetBuildInnerRepo` is not set in VMR, it is only set in repo's inner build. Is that going to change, should we switch to checking `'$(DotNetBuildPhase)' == '$(InnerRepo)'`?

File: src/Microsoft.DotNet.Arcade.Sdk/tools/SourceBuild/SourceBuildArcadeBuild.targets (hunk 14 above)
Comment (NikolaMilosavljevic) [reply src/Microsoft.DotNet.Arcade.Sdk/tools/SourceBuild/SourceBuildArcadeBuild.targets:85]: This is needed so we avoid publishing of symbols to symbol server: https://github.com/dotnet/arcade/blob/a822ee27362347b6bb0d7e7878e0b6961a69a680/src/Microsoft.DotNet.Arcade.Sdk/tools/Build.proj#L99-L106

I think the comment should be updated here.

File: src/Microsoft.DotNet.Arcade.Sdk/tools/Publish.proj (hunk 15 above)
Comment (mmitche) [reply src/Microsoft.DotNet.Arcade.Sdk/tools/Publish.proj:29]: `DotNetBuildInnerRepo` is set in VMR builds, when building the inner repo. It's also set in repo builds (not VMR) when building the inner repo. It just says that the inner repo build is in progress. The way the switches are done is that a flag (e.g. DotNetBuildRepo) is set for the phase it represents (outer repo build) and any phases it contains (e.g. inner repo). DotNetBuildPhase represents the exclusive phase that happens to be executing right now.

What your above condition is saying is "publish to local storage when building in context of the VMR, when building the inner repo". I think you probably meant this, unless you want repo source builds (non-VMR) to use the new publishing infra. Assuming you meant it as written, I would rephrase it as:

`<PropertyGroup Condition="'$(DotNetBuildPhase)' == 'Inner' and '$(DotNetBuildOrchestrator)' == 'true'">` with a comment indicating the reason why inner builds outside of the VMR are left out.

File: src/Microsoft.DotNet.Arcade.Sdk/tools/SourceBuild/SourceBuildArcadeBuild.targets (hunk 14 above)
Comment (mmitche) [reply src/Microsoft.DotNet.Arcade.Sdk/tools/SourceBuild/SourceBuildArcadeBuild.targets:85]: Yes, this is correct. IMO the `DotNetPublishUsingPipelines` switch should be removed at some point, or at least set to true as the default. There are very very few repos that don't use this today.

File: src/Microsoft.DotNet.Arcade.Sdk/tools/Publish.proj (hunk 15 above)
Comment (NikolaMilosavljevic) [reply src/Microsoft.DotNet.Arcade.Sdk/tools/Publish.proj]: Fixed with https://github.com/dotnet/arcade/pull/14559/commits/dd0cef5f21397ad1d7ff4e52be3359fc77d2dd1d

**Review by ViktorHofer — APPROVED** [review]

File: src/Microsoft.DotNet.VersionTools/lib/src/Automation/NupkgInfoFactory.cs (hunk 9 above)
Comment (ViktorHofer) [reply src/Microsoft.DotNet.VersionTools/lib/src/Automation/NupkgInfoFactory.cs:48]: Addressed via https://github.com/dotnet/arcade/pull/14589

File: src/Microsoft.DotNet.VersionTools/tools/Microsoft.DotNet.VersionTools.Cli/Microsoft.DotNet.VersionTools.Cli.csproj (hunk 1 above)
Comment (ViktorHofer) [reply src/Microsoft.DotNet.VersionTools/tools/Microsoft.DotNet.VersionTools.Cli/Microsoft.DotNet.VersionTools.Cli.csproj]: Addressed via https://github.com/dotnet/arcade/pull/14588

File: src/Microsoft.DotNet.VersionTools/lib/src/Automation/NupkgInfo.cs (hunk 8 above)
Comment (ViktorHofer) [reply src/Microsoft.DotNet.VersionTools/lib/src/Automation/NupkgInfo.cs:47]: Addressed via https://github.com/dotnet/arcade/pull/14589
